sympy/matrices/common.py

```diff
@@ -2540,3 +2555,75 @@ def classof(A, B):
             return A.__class__
 
     raise TypeError("Incompatible classes %s, %s" % (A.__class__, B.__class__))
+
+
+def dotprodsimp(a, b, simplify=True):
+    """Sum-of-products with optional intermediate product simplification
+    targeted at the kind of blowup that occurs during summation of products.
+    Intended to reduce expression blowup during matrix multiplication or other
+    similar operations.
+
+    Parameters
+    ==========
+
+    a, b : iterable
+        These will be multiplied then summed together either normally or
+        using simplification on the intermediate products and cancelling at
+        the end according to the 'simplify' flag. The elements must already be
+        sympyfied and the sequences need not be of the same length, the shorter
+        will be used.
+
+    simplify : bool
+        When set intermediate and final simplification will be used, not set
+        will indicate a normal sum of products.
+    """
+
+    itra = iter(a)
+    itrb = iter(b)
+
+    try: # check for zero length
+        prod = next(itra)*next(itrb)
+    except StopIteration:
+        return S.Zero
+
+    # simple non-simplified sum of products
+    if not simplify:
+        return reduce (lambda e, z: e + z[0]*z[1], zip(itra, itrb), prod)
+
+    # part 1, the expanded summation
+    _expand = getattr(prod, 'expand', None)
+    expr    = _expand(power_base=False, power_exp=False, log=False, \
+            multinomial=False, basic=False) if _expand else prod
```

Comment:
This code (for _expand) is duplicated below. It would be better to use a function